crates/clvm-utils/src/tree_hash.rs

```diff
@@ -307,6 +308,79 @@ pub fn tree_hash_cached(a: &Allocator, node: NodePtr, cache: &mut TreeCache) ->
     hashes[0]
 }
 
+fn subtract_cost(cost_left: &mut Cost, subtract: Cost) -> Result<(), usize> {
+    if subtract > *cost_left {
+        Err(0)
+    } else {
+        *cost_left -= subtract;
+        Ok(())
+    }
+}
+
+pub fn tree_hash_cached_costed(
+    a: &Allocator,
+    node: NodePtr,
+    cache: &mut TreeCache,
+    cost_left: &mut u64,
+    cost_per_call: u32,
+    cost_per_byte: u32,
+) -> Result<TreeHash, usize> {
+    cache.visit_tree(a, node);
+
+    let mut hashes = Vec::new();
+    let mut ops = vec![TreeOp::SExp(node)];
+
+    while let Some(op) = ops.pop() {
+        subtract_cost(cost_left, cost_per_call.into())?;
+        match op {
+            TreeOp::SExp(node) => match a.node(node) {
+                NodeVisitor::Buffer(bytes) => {
+                    subtract_cost(cost_left, (cost_per_byte * bytes.len() as u32).into())?;
```

Comment:
```suggestion
                    subtract_cost(cost_left, u64::from(cost_per_byte) * u64::from(bytes.len()))?;
```
the `as` operator isn't as safe as `::from()` since it will silently truncate values. `Cost` is u64, so `u32` doesn't seem like the right type